utils/i18nUtils.js

```diff
@@ -0,0 +1,118 @@
+import { match } from "@formatjs/intl-localematcher";
+import { shouldPolyfill } from "@formatjs/intl-locale/should-polyfill";
+
+// Update this to the default locale for the app
+const DEFAULT_LOCALE = "en";
+// Update this to the supported locales contained in the `locales` folder
+const SUPPORTED_APP_LOCALES = ["en", "fr"];
+
+// TODO: Export polaris supported locales from the Polaris package
+const SUPPORTED_POLARIS_LOCALES = [
+  "cs",
+  "da",
+  "de",
+  "en",
+  "es",
+  "fi",
+  "fr",
+  "it",
+  "ja",
+  "ko",
+  "nb",
+  "nl",
+  "pl",
+  "pt-BR",
+  "pt-PT",
+  "sv",
+  "th",
+  "tr",
+  "vi",
+  "zh-CN",
+  "zh-TW",
+];
+
+let _userLocale, _polarisTranslations;
+
+export function getUserLocale() {
+  if (_userLocale) {
+    return _userLocale;
+  }
+  const url = new URL(window.location.href);
+  const locale = url.searchParams.get("locale") || DEFAULT_LOCALE;
+  _userLocale = match([locale], SUPPORTED_APP_LOCALES, DEFAULT_LOCALE);
+  return _userLocale;
+}
+
+export function getPolarisTranslations() {
+  return _polarisTranslations;
+}
+
+export async function initI18n() {
+  if (shouldPolyfill()) {
+    await import("@formatjs/intl-locale/polyfill");
+  }
+  await Promise.all([initI18next(), fetchPolarisTranslations()]);
+}
+
+async function initI18next() {
+  return await i18next
+    .use(initReactI18next)
+    .use(localResourcesToBackend())
+    .init({
+      debug: process.env.NODE_ENV === "development",
+      lng: getUserLocale(),
+      fallbackLng: DEFAULT_LOCALE,
+      supportedLngs: SUPPORTED_APP_LOCALES,
+      interpolation: {
+        // React escapes values by default
+        escapeValue: false,
+      },
+      react: {
+        // Wait for the locales to be loaded before rendering the app
+        // instead of using a Suspense component
+        useSuspense: false,
+      },
+    });
+}
+
+function localResourcesToBackend() {
+  return resourcesToBackend(async (locale, _namespace, callback) => {
+    import(`../locales/${locale}.json`)
+      .then(({ default: resources }) => {
+        callback(null, resources);
+      })
+      .catch((error) => {
+        console.error(`Error loading translations for ${locale}:`, error);
+        callback(error, null);
+      });
+  });
+}
+
+async function fetchPolarisTranslations() {
+  if (_polarisTranslations) {
+    return _polarisTranslations;
+  }
+  // Get the closest matching default locale supported by Polaris
+  const defaultPolarisLocale = match(
+    [DEFAULT_LOCALE],
+    SUPPORTED_POLARIS_LOCALES,
+    "en"
+  );
+  // Get the closest matching user locale supported by Polaris
+  const polarisLocale = match(
+    [getUserLocale()],
+    SUPPORTED_POLARIS_LOCALES,
+    defaultPolarisLocale
+  );
+  _polarisTranslations = await loadPolarisTranslations(polarisLocale);
+  return _polarisTranslations;
+}
+
+async function loadPolarisTranslations(locale) {
+  return (
+    await import(`../node_modules/@shopify/polaris/locales/${locale}.json`)
```

Comment:
My app is build by ESBuild and I am having the same problem here. Once the JS is loaded by the browser, it tries to get the polaris' locale making a `GET` request to `/node_modules/@shopify/polaris/locales/en.json`. It gets a 404 error.

This problem happens because we cannot dynamically import libraries using a variable inside the path. The complire does not know the path (as it uses variables) so it does not import locales. 

Here is the code I use before a better solution is found:
```javascript
async function loadPolarisTranslations(locale) {
  let lib;
  if (locale == 'en') {
    lib = import(`../node_modules/@shopify/polaris/locales/en.json`)
  } else if (locale == 'fr') {
    lib = import(`../node_modules/@shopify/polaris/locales/fr.json`)
  }
  return (
    await lib
  ).default;
}
```